### PR TITLE
[Merged by Bors] - feat: inverse of a 1×1 matrix

### PR DESCRIPTION
**Pull request by eric-wieser**
The statement is a little strange, so it comes with a docstring to justify it.

From https://github.com/eric-wieser/lean-matrix-cookbook

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean [hunk 1]
```diff
@@ -607,6 +611,24 @@ theorem inv_diagonal (v : n → α) : (diagonal v)⁻¹ = diagonal (Ring.inverse
...
+/-- The inverse of a `1x1` or `0x0` matrix is always diagonal.
+
+While we could write this as `of fun _ _ => Ring.inverse (A default default)` on the RHS, this is
+less useful because:
+
+* It wouldn't work for `0x0` matrices.
+* More things are true about diagonal matrices than constant matrices, and so more lemmas exist.
+
+`Matrix.diagonal_unique` can be used to reach this form, while `Ring.inverse_eq_inv` can be used
+to replace `Ring.inverse` with `⁻¹`.
```
Comment (b-mehta) [inline Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean:623]: I agree inverse_eq_inv can be used to replace the inverse with an inv, but isn't the lemma clearer if you do the replacing here?

File: Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean [hunk 2]
```diff
@@ -607,6 +611,24 @@ theorem inv_diagonal (v : n → α) : (diagonal v)⁻¹ = diagonal (Ring.inverse
...
+
+* It wouldn't work for 0×0 matrices.
+* More things are true about diagonal matrices than constant matrices, and so more lemmas exist.
+
+`Matrix.diagonal_unique` can be used to reach this form, while `Ring.inverse_eq_inv` can be used
+to replace `Ring.inverse` with `⁻¹`.
+-/
+@[simp]
+theorem inv_subsingleton [Subsingleton m] [Fintype m] [DecidableEq m] (A : Matrix m m α) :
+    A⁻¹ = diagonal fun i => Ring.inverse (A i i) := by
```
Comment (b-mehta) [inline Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean:627]: I'm not sure this should be `simp` - it'll fire any time there's an inverse matrix around, and the search for `Subsingleton m` could be expensive

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean:627]: `Matrix.det_unique` is already simp; are the considerations for `Unique` different to those for `subsingleton?

**Comment by eric-wieser** [comment]
!bench

File: Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean (hunk 2 above)
Comment (b-mehta) [reply Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean:627]: Potentially, but I'd guess the larger problem is that `det` shows up a little less often than `⁻¹` on matrices

**Comment by b-mehta** [comment]
Thanks! 

bors merge

**Pull request closed without merging**